### ISSUE TITLE
[Discussion] Diffuser Framework vs WebUI Extension

### ISSUE THREAD

**Issue by plcpinho** [haofanwang/ControlNet-for-Diffusers#10]
Hello @haofanwang,

I am trying to replicate Mikubill Transfer Control. In this way, initially I followed [this guide](https://github.com/lllyasviel/ControlNet/discussions/12) and after comparing the AnyV3_Canny_Model.pth created to the WebUI (using same prompt, seed etc.), I realized that they are not the same... is it normal? Please, check the differences below:
![Blank 2 Grids Collage (1)](https://user-images.githubusercontent.com/25768651/221198727-8001fce2-f4e1-404b-b0e3-20b2bc99310b.png)
In addition, I am trying to save the merged files when using [Mikubill repo](https://github.com/Mikubill/sd-webui-controlnet/), tried to add the following line in cldm.py (in the end of the `PlugableControlModel init()`):
```torch.save(final_state_dict, './control_any3_openpose.pth')```
But i don't think it is correct... do you have any thoughts on this?

**Comment by haofanwang** [comment]
(1) How does **AnyV3_Canny_Model.pth** created in WebUI? Isn't it using [the method](https://github.com/lllyasviel/ControlNet/discussions/12) described in controlnet?

(2) What do you mean by **not the same**. The weights have different keys or values, or both?

(3) Can you provide more details about what you want to? What is your base-model and ControNet (what condition)? If possible, could you attach the input and control image (pose, depth or else), so that I can have a try.

@plcpinho

**Comment by plcpinho** [comment]
(1) I didn't create **AnyV3_Canny_Model.pth** via the WebUI. This is one of my goals, but I don't think it is possible because as I've understood Mikubill is not merging the models like your doing, he's doing it "on the fly" - didn't understand exactly how. The comparison above is **AnyV3_Canny_Model.pth** created via [this guide](https://github.com/lllyasviel/ControlNet/discussions/12) vs Mikubill WebUI image generation;

(2) As both methods (Diffusers and Mikubill WebUI) try to achieve the same result (CustomSD15+ControlNet), I thought about comparing both methods and checking if they generate the same or similar images (using the same prompt, seed etc.). And the result was the above (different images).

(3) To be clearer, my goal is to create the best CustomSD15+ControlNet model (e.g. **AnyV3_Canny_Model.pt**) possible. And I found that Mikubill WebUI is generating better images if compared to Diffusers merging methods. Does it make sense or is this a wrong statement?

_PS.:
- CustomSD15 models I am referring to any SD15 model, such as AnythingV3, RealisticVision etc. 
- ControlNet models I am referring to SD15Canny, SD15OpenPose etc._ 

@haofanwang Thank you much for replying the thread!

**Comment by plcpinho** [comment]
I am trying to analyze Mikubill code in order to create a **AnyV3_Canny_Model.pth** file and then compare the weights of both methods (Mikubill vs `tool_transfer_control.py`), but I am not succeeding, because as I said I suspect there is no model merge. 😓

**Comment by haofanwang** [comment]
I totally agree with you. I also fail to figure out how does webui work on the fly. That is why I make this project.

One possible reason is that `tool_transfer_control.py` may miss out something, and is not fully converted. I'm not sure whether the released ControlNet weights are completely independent of UNet, or as mentioned in ControlNet repo, it may also train some layers into UNet. In such a case, we only load part of weights.

Anyway, thanks for your finding, I'm also interested in converting on the fly, we can keep updated. @plcpinho

**Comment by plcpinho** [comment]
@haofanwang I just found something very interesting! This [guide](https://github.com/lllyasviel/ControlNet/discussions/12) was written by Illyasviel 2 weeks ago, last week a [thread](https://github.com/Mikubill/sd-webui-controlnet/issues/73) was created in which **Illyasviel, Mikubill and Kohya-ss** discussed about Transfer Control implementations. In one part of the conversation, Kohya-ss made a [pull request](https://github.com/Mikubill/sd-webui-controlnet/pull/80) for a method (which is btw the currently method used in Mikubill's repo) of Transfer Control. In Kohya-ss' words: "extract_controlnet_diff.py makes the difference and save the state_dict with key difference as a marker, and cldm.py handles it on the fly.".

I believe the answer to our questions is in this thread and in this PR mentioned above!

**Comment by ghpkishore** [comment]
@plcpinho can we for testing, use the Mikubill repo and get the canny model and then use that ? Should that work? I fully didn't understand how to solve this, except that I too ran into an error when trying to create canny edge based inpainting output. I would appreciate if you have can give specific steps through which we can use the new control models in our inpainting pipeline.

**Comment by haofanwang** [comment]
@plcpinho I'm glad to know this! I can work on it based on your info. If you are interested in, PR is very very welcome.

**Comment by haofanwang** [comment]
Will dive into https://github.com/Mikubill/sd-webui-controlnet/pull/80 and https://github.com/Mikubill/sd-webui-controlnet/issues/73. If anyone is willing to help, please let me know.

**Comment by haofanwang** [comment]
I don't find anything difference between `merging beforehand` and `merging on the fly`. They actually do the same. Still not clear what leads to such a difference, will spend more time on investing this.

- **merging beforehand**: This is what [ControlNet/discussions/12](https://github.com/lllyasviel/ControlNet/discussions/12) does, so do our step (1).

-  **merging on the fly**: This is what [sd-webui-controlnet/pull/80](https://github.com/Mikubill/sd-webui-controlnet/pull/80) does. If you look at the code in [cldm.py](https://github.com/Mikubill/sd-webui-controlnet/blob/72b015b1ad352585bd009fee1974550cc8ecf2dd/scripts/cldm.py#L87), they just conduct merging as the same as ControlNet does in [tool_transfer_control.py](https://github.com/lllyasviel/ControlNet/blob/1234076190105dc6541de3bc2aa7b3d24f8e3469/tool_transfer_control.py#L50). [extract_controlnet_diff.py](https://github.com/Mikubill/sd-webui-controlnet/blob/main/extract_controlnet_diff.py) and [extract_controlnet.py](https://github.com/Mikubill/sd-webui-controlnet/blob/main/extract_controlnet.py) do not matter here, the first one just calculates the offset (model.control.unet - sd.unet) beforehand.

For now, I have no plan to add **merging on the fly** in this repo, as we more care about using in diffusers. Our goal is to load the new model via diffusers function `from_pretrained()`.

**Comment by plcpinho** [comment]
Thank you for your reply, @haofanwang!

Technically **merging beforehand** and **on the fly** do the same. However, I am not sure if the models generate the same/similar images. I am going to do some more tests today! Btw, you said in [this thread](https://github.com/Mikubill/sd-webui-controlnet/issues/73) that the model is actually being merged in both methods, do you know how to **save the merged model** in [sd-webui-controlnet](https://github.com/Mikubill/sd-webui-controlnet)? Because, if we have both models saved, it is easy to compare the merging **beforehand** vs **on the fly**. I tried to add the following line in [clmd.py](https://github.com/Mikubill/sd-webui-controlnet/blob/72b015b1ad352585bd009fee1974550cc8ecf2dd/scripts/cldm.py#L106):

```python
# Does not work... saves a ~700KB file
torch.save(state_dict, './merged_model.pth')
```

About the [cldm.py](https://github.com/Mikubill/sd-webui-controlnet/blob/72b015b1ad352585bd009fee1974550cc8ecf2dd/scripts/cldm.py#L67-L102), I am also not sure if this code is actually running, because if look the "if" statement (line 66) there is never a "k" item starting with "control_model.". So, the operations you mentioned above, during my tests, never gets called. You can see that by adding some print() and running ControlNet via the WebUI... 🤔

**Comment by haofanwang** [comment]
It's kind of weird, let's follow up.

**Comment by ghpkishore** [comment]
@paulo-coronado any updates on how the canny model is differing? Also, I would appreciate if you can share more on if it is only the canny model which is not the same as Mikubil webui or if others are also differing.

**Comment by paulo-coronado** [comment]
Hey, @ghpkishore! I don't know why it is differing, but I can tell you that merging using [this guide](https://github.com/lllyasviel/ControlNet/discussions/12) works! Although it generates different results, the images are still great! I might have done something wrong to have generated the above example...

**Comment by ghpkishore** [comment]
Thanks. Do you know how I can use safe tensors and convert that into normal PyTorch bin and use that.  I am running into some error when I try to use the control net canny edge model already made which say that the weights are not initialised.

Tried for couple of hours and gave up

**Comment by paulo-coronado** [comment]
Do you have Discord, @ghpkishore? Send me your profile so me can chat there :)

**Comment by ghpkishore** [comment]
@paulo-coronado It is the same. @ghpkishore

**Comment by paulo-coronado** [comment]
@ghpkishore The Discord username is something like @ghpkishore#0000

**Comment by ghpkishore** [comment]
@paulo-coronado ghpkishore#4438
